authentication/account/worker/user_deactivation_notification_worker.go

```diff
@@ -0,0 +1,116 @@
+package worker
+
+import (
+	"context"
+	"time"
+
+	"cirello.io/pglock"
+
+	"github.com/fabric8-services/fabric8-auth/application"
+	"github.com/fabric8-services/fabric8-auth/log"
+	worker "github.com/fabric8-services/fabric8-auth/worker/repository"
+)
+
+// UserDeactivationNotificationWorker the interface for the User Deactivation Worker,
+// which takes care of deactivating accounts of inactive users who were previously notified but did not come back afterwards.
+type UserDeactivationNotificationWorker interface {
+	Start(freq time.Duration)
+	Stop()
+}
+
+// NewUserDeactivationNotificationWorker returns a new UserDeactivationNotificationWorker
+func NewUserDeactivationNotificationWorker(ctx context.Context, app application.Application) UserDeactivationNotificationWorker {
+	return &userDeactivationNotificationWorker{
+		ctx:   ctx,
+		app:   app,
+		owner: worker.GetLockOwner(ctx),
+	}
+}
+
+type userDeactivationNotificationWorker struct {
+	ctx    context.Context
+	app    application.Application
+	lock   *pglock.Lock
+	ticker *time.Ticker
+	stopCh chan bool
+	owner  string
+}
+
+// Start starts the worker with the given timer
+func (w *userDeactivationNotificationWorker) Start(freq time.Duration) {
```

Comment:
We can't guarantee that Start() will be always called. So, we have to come up with some health check. For example we could write a timestamp of the last worker cycle and check it when starting a worker. But we should also take into account rolling updates.
Take a look at our tenant update mechanism which deals with the same problem. @MatousJobanek can help you with that.

Reply:
@alexeykazakov the `Start(time.Duration)` func will be called in the `main.go` during the bootstrap, so I'm not sure why you say that we can't guarantee it will always be called. Could you elaborate, please?
More importantly, we must ensure that the lock will be released during shutdown, ie, we must have a "graceful shutdown" function to handle that..

Reply:
Oh, did I say Start()? :)
I meant Stop() of course! :)
We can’t guarantee it’s always called. Even defer is not 100% guaranty.

Reply:
well, I hope that a graceful shutdown hook can at least release the lock in a proper manner, which would work for the rolling updates. But that would not work for the case when the process is "hard killed", indeed.

Reply:
I guess that this should be handled by `pglock` library - the [heartbeat frequency](https://github.com/cirello-io/pglock#send-automatic-heartbeats) should be exactly for this, right? However, the question is if it is handled correctly. 

In addition, could it happen that when the pod that is running the worker is killed, and a new one is started quicker than the lock is released? This scenario could be actually solved by a `time.Sleep(lease_duration)` at the beginning of the goroutine that starts the worker. Or, more gracefully,  sleep for that time when the acquiring of the lock fails and then try it one more time. If the second attempt fails as well, then stop trying it.

Reply:
This is a good point @MatousJobanek.
Heartbeat thing looks exactly what we need here. Just let make sure we use it correctly with all timeouts set correctly, etc.

> Or, more gracefully, sleep for that time when the acquiring of the lock fails and then try it one more time. If the second attempt fails as well, then stop trying it.

+1. Just need to make sure that the sleep is longer than the Heartbeat frequency.

Reply:
ok, so with https://github.com/fabric8-services/fabric8-auth/pull/812/commits/e7b23af3ca548b4a2b2c7a936c2951950d2d425a, I assume that things are a little simpler: if a worker cannot acquire the lock, then it will try again, at the next cycle. The question remains for the case where the pod is killed while the worker acquired the lock and is still working. I'll see how the heartbeat can help in that case (with some tests, if possible)